src/Widget/Widget.php

```diff
@@ -9,129 +9,76 @@
 use Yiisoft\View\WebView;
 use Yiisoft\Widget\Event\AfterRun;
 use Yiisoft\Widget\Event\BeforeRun;
+use Yiisoft\Widget\Exception\InvalidConfigException;
 
 /**
  * Widget is the base class for widgets.
  *
  * For more details and usage information on Widget, see the [guide article on widgets](guide:structure-widgets).
  */
-class Widget implements ViewContextInterface
+class Widget
 {
     /**
-     * @var EventDispatcherInterface event handler.
+     * @var EventDispatcherInterface $eventDispatcher
      */
-    protected static $eventDispatcher;
+    private $eventDispatcher;
 
     /**
-     * The widgets that are currently being rendered (not ended). This property is maintained by {@see static::begin()}
-     * and {@see static::end()} methods.
+     * The widgets that are currently being rendered (not ended). This property is maintained by {@see begin()} and
+     * {@see end} methods.
      *
      * @var Widget[] $stack
      */
-    protected static $stack;
+    private $stack;
 
-    /**
-     * @var WebView $view
-     */
-    protected static $webView;
-
-    /**
-     * @var Widget $widget
-     */
-    protected static $widget;
-
-    public function __construct(EventDispatcherInterface $eventDispatcher, WebView $webView)
+    public function __construct(WebView $webView)
     {
-        self::$eventDispatcher = $eventDispatcher;
-        self::$webView = $webView;
+        $this->eventDispatcher = $webView->getEventDispatcher();
     }
 
     /**
-     * Begins a widget.
-     *
-     * This method creates an instance of the calling class. It will apply the configuration to the created instance.
-     * A matching {@see end()} call should be called later. As some widgets may use output buffering, the {@see end()}
-     * call should be made in the same view to avoid breaking the nesting of output buffers.
-     *
-     * @return Widget the newly created widget instance.
+     * Begin the rendering of content.
      *
-     * {@see end()}
+     * @return Widget
      */
-    public static function begin(): Widget
+    public function begin(): Widget
     {
-        $widget = new static(self::$eventDispatcher, self::$webView);
+        $this->stack[] = $this;
```

Comment:
What is the point of keeping stack at widget instance level? It will always contain one element - the widget itself.

Reply:
It is maintained to be able to check that each begin() has an end() basically. Another type of check could be implemented.

Reply:
> It is maintained to be able to check that each begin() has an end() basically.

Where is this check? AFAIK you can still call `begin()` and `end()` for separate widgets in incorrect order. And call `begin()` without `end()`.

Reply:
It is correct but throw the exception, you can review the tests, now that you are glad to suggest any improvement is welcome.

Reply:
Agreeing with @rob006, this property made sense when it was static and shared by all widgets.

This is not the case anymore, there is no "stack" here, and as far as I read, it will never contain more than an element, so having an array doesn't make sense.

Reply:
@rob006 Now I understood his point of view, and he is right.

Reply:
Stack could/should be implemented at view level.

Reply:
@rob006 if stack was implemented at View level, the View should be hard linked to Widgets. I think that it would be better that stack was inside Widget, so View should not know anything about Widget.

Reply:
It is correct, widget does not depend on ViewContextInterface, it can even be separated from the view as an additional component.

Reply:
@FabrizioCaldarelli Why it should not know anything about widget? Widgets are usually rendered inside of view, using view for controlling them looks quite natural for me. 

Having a stack at widget level is essentially useless, since each widget will have own stack. Main purpose of stack is avoiding situations like:

```php
$widgetA->begin();
$wigdetB->begin();
$widgetA->end();
$widgetB->end();
```

Stack worked as a guard to prevent such errors and simplify syntax (you don't need to keep instance, `Widget::end()` will find correct instance based on stack). None of this works with current implementation.

Reply:
Thank you @rob006 for your comment, this implementation is right that does not solve that problem, I will think of a better solution.

Reply:
@rob006 Because widgets are not a "property" of a View, but they are components used in a View.
For this reason, View should know anything about Widgets.
Otherwise, for every new type of objects used in a View, the View should be responsible of this, keeping a reference inside a View.
I'm agree with keeping stack handle inside the Widget.

Reply:
> Because widgets are not a "property" of a View, but they are components used in a View.

And so what? :) Some syntactic sugar for rendering widget will not hurt. You're just making things harder/impossible to implement for the sake of some arbitrary "should know nothing" rule.

Reply:
> > Because widgets are not a "property" of a View, but they are components used in a View.
> 
> And so what? :) Some syntactic sugar for rendering widget will not hurt. You're just making things harder/impossible to implement for the sake of some arbitrary "should know nothing" rule.

Not for the sake, but because it was already implemented in the past (yii2 widget version).

So I'd try to take the same way to solve this trouble.

Reply:
> Not for the sake, but because it was already implemented in the past (yii2 widget version).

Yii 2 used static property to keep stack, which should be unacceptable in Yii 3 - there should be no global static state for components, it makes them unpredictable.